### PR TITLE
Add system prompt template

### PR DESCRIPTION
**Pull request by leehuwuj**
image

### CODE REVIEW

File: src/ragapp/config/agents.yaml [hunk 1]
```diff
@@ -1,27 +1,37 @@
...
+  goal: To help answer questions using the provided knowledge.
   name: Default
-  role: Expert in answering questions using the provided knowledge
+  role: Assistant
   system_prompt: You are a helpful assistant.
+  system_prompt_template: null
   tools:
     DuckDuckGo:
       config: {}
+      custom_prompt: null
```
Comment (leehuwuj) [inline src/ragapp/config/agents.yaml:12]: Use a shared config for tools to avoid duplicated config

File: src/ragapp/admin-ui/client/agent.ts [hunk 2]
```diff
@@ -40,12 +40,12 @@ export const AgentConfigSchema = z.object({
...
+  backstory: z.string(),
+  goal: z.string(),
+  system_prompt: z.string().nullable(),
+  system_prompt_template: z.string().nullable(),
   tools: ToolsSchema,
-  created_at: z
-    .string()
-    .or(z.date())
-    .transform((val) => new Date(val)),
+  created_at: z.number(), // Change to number
```
Comment (marcusschiesser) [inline src/ragapp/admin-ui/client/agent.ts:47]: better tell what this number means
```suggestion
  created_at: z.number(), // Change to number
```

File: src/ragapp/admin-ui/sections/config/agent.tsx [hunk 3]
```diff
@@ -60,7 +60,7 @@ export const AgentConfig = () => {
       onMutate: () => setIsSubmitting(true),
```
Comment (marcusschiesser) [inline src/ragapp/admin-ui/sections/config/agent.tsx:60]: some of the comments in this file are helpful

File: src/ragapp/backend/agents/orchestrator.py [hunk 4]
```diff
@@ -41,11 +41,17 @@ def get_agents(
             for tool_name, tool_config in agent_tools_config
             if tool_config.enabled
         ]
+        system_prompt = agent_config.get_system_prompt()
+        # Role is an information to introduce the agent to the orchestrator
+        # construct a custom "role" from the original role and goal for better orchestration
+        refined_agent_role = (
+            f"{agent_config.role}\n and its goals are {agent_config.goal}"
+        )
```
Comment (marcusschiesser) [inline src/ragapp/backend/agents/orchestrator.py:49]:
```suggestion
        # The orchestrator uses "role" to select the right agent for a task
        # construct an "agent_selector" from the user defined role and goal for better orchestration
        agent_selector = (
            f"{agent_config.role}\n and its goals are {agent_config.goal}"
        )
```

File: src/ragapp/backend/agents/orchestrator.py [hunk 5]
```diff
@@ -41,11 +41,17 @@ def get_agents(
...
         agents.append(
             FunctionCallingAgent(
                 name=agent_config.name,
-                role=agent_config.role,
-                system_prompt=agent_config.system_prompt,
+                role=refined_agent_role,
```
Comment (marcusschiesser) [inline src/ragapp/backend/agents/orchestrator.py:53]:
```suggestion
                role= agent_selector,
```

File: src/ragapp/backend/agents/orchestrator.py (hunk 4 above)
Comment (marcusschiesser) [reply src/ragapp/backend/agents/orchestrator.py:49]: @leehuwuj This is a potential problem if the user doesn't select goal and role, the selection won't work

Suggestions:
1. we make role and goal required parameters
2. if the user selects a custom_prompt, he should also select the agent selector, defaults to "{role} and its goals are {goal}"

Comment (marcusschiesser) [reply src/ragapp/backend/agents/orchestrator.py:49]: BTW, isn't the original string: `The agent is an {role}`?

File: src/ragapp/admin-ui/sections/config/agents/AgentTabContent.tsx [hunk 6]
```diff
@@ -43,73 +54,185 @@ export const AgentTabContent = ({
...
                         {...field}
                         onBlur={handleInputBlur}
-                        placeholder="Enter agent name"
+                        rows={3}
+                        value={field.value || ""}
                       />
                     </FormControl>
                     <FormDescription>
-                      Unique name to identify the agent.
+                      {`This prompt will be used to generate the system prompt for the agent. Use the {role}, {backstory}, and {goal} variables to use the values from the agent config.`}
```
Comment (marcusschiesser) [inline src/ragapp/admin-ui/sections/config/agents/AgentTabContent.tsx:239]:
```suggestion
                      {`This prompt will be used to generate the system prompt for the agent. You can use the {role}, {backstory}, and {goal} variables from the agent's configuration.`}
```

File: src/ragapp/admin-ui/client/agent.ts [hunk 7]
```diff
@@ -39,13 +39,16 @@ export const ToolsSchema = z.object({
 export const AgentConfigSchema = z.object({
   agent_id: z.string(),
   name: z.string(),
-  role: z.string(),
-  system_prompt: z.string(),
+  role: z.string().trim().min(1, {
+    message: "Role is required. It's used to identify the agent.",
```
Comment (marcusschiesser) [inline src/ragapp/admin-ui/client/agent.ts:43]:
```suggestion
    message: "Role is required to select the right agent for a task.",
```

File: src/ragapp/admin-ui/client/agent.ts [hunk 8]
```diff
@@ -39,13 +39,16 @@ export const ToolsSchema = z.object({
...
+  }),
+  goal: z.string().min(1, {
+    message: "Goal is required. It's used to define the agent's purpose.",
```
Comment (marcusschiesser) [inline src/ragapp/admin-ui/client/agent.ts:46]:
```suggestion
    message: "Goal is required to select the right agent for a task.",
```

File: src/ragapp/admin-ui/sections/config/agents/AgentTabContent.tsx [hunk 9]
```diff
@@ -43,73 +73,197 @@ export const AgentTabContent = ({
...
+                      <Tooltip>
+                        <TooltipTrigger asChild>
+                          <span tabIndex={-1}>
+                            <InfoIcon className="h-4 w-4 text-muted-foreground cursor-help" />
+                          </span>
+                        </TooltipTrigger>
+                        <TooltipContent>
+                          <p>
+                            Define the agent&apos;s primary objective. Useful
+                            for deliver right tasks to the agent.
```
Comment (marcusschiesser) [inline src/ragapp/admin-ui/sections/config/agents/AgentTabContent.tsx:215]:
```suggestion
                            Define the agent&apos;s primary objective. Required to
                            select the right agent for a task.
```

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**